src/core/Grappa.sol

```diff
@@ -87,8 +91,36 @@ contract Grappa is Ownable, ReentrancyGuard {
     event MarginEngineRegistered(address engine, uint8 id);
     event OracleRegistered(address oracle, uint8 id);
 
-    constructor(address _optionToken) {
+    /*///////////////////////////////////////////////////////////////
+                Constructor for implementation Contract
+    //////////////////////////////////////////////////////////////*/
+
+    /// @dev set immutables in constructor
+    /// @dev also set the implemention contract to initialized = true
+    constructor(address _optionToken) initializer {
         optionToken = IOptionToken(_optionToken);
+        deployTimestamp = block.timestamp;
+    }
+
+    /*///////////////////////////////////////////////////////////////
+                            Initializer
+    //////////////////////////////////////////////////////////////*/
+
+    function initialize() external initializer {
+        __Ownable_init();
+    }
+
+    /*///////////////////////////////////////////////////////////////
+                    Override Upgrade Permission
+    //////////////////////////////////////////////////////////////*/
+
+    /**
+     * @dev Upgradable by the owner, can only be upgraded within 365 days after deployment
+     **/
+    function _authorizeUpgrade(
+        address /*newImplementation*/
+    ) internal view override onlyOwner {
+        if (block.timestamp > deployTimestamp + 365 days) revert GP_NotUpgradable();
```

Comment:
why do you want to limit it in this way?

Reply:
Is this method immune to upgrades? From what i am reading it is not, so one could upgrade grappa and remove this limit.

Reply:
yeah, it's true, just like you said, we can bypass this and make the next version not upgradable.
But if we really leave it for 1 year, it will become not upgradable, so if that day come, people can trust the contract to not change.

(IMO adding to the proxy directly is too dangerous, I really want to avoid changing the proxy code at all)

Reply:
To clarify, i am proponent of removing the expiration entirely. Feels like a ticking time bomb. If the concern is around sufficient notice around upgrades then we should add a timelock so that anyone who is interested could listen for the event.

Reply:
I certainly dont want to change anything in the proxy code.

Reply:
decision: remove for now to unblock other PRs